<!doctype html>
<html lang="en">
<head>
<title>2017-06-19-tgen</title>
<meta name="generator" content="Etherpad">
<meta name="author" content="Etherpad">
<meta name="changedby" content="Etherpad">
<meta charset="utf-8">
<style>
* {
  font-family: arial, sans-serif;
  font-size: 13px;
  line-height: 17px;
}
ul.indent {
  list-style-type: none;
}
ol {
  list-style-type: none;
  padding-left: 0;
}
body > ol {
  counter-reset: first second third fourth fifth sixth seventh eigth ninth tenth eleventh twelth thirteenth fourteenth fifteenth sixteenth;
}
ol > li:before {
  content: counter(first) ". ";
  counter-increment: first;
}
ol > ol > li:before {
  content: counter(first) "." counter(second) ". ";
  counter-increment: second;
}
ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) ". ";
  counter-increment: third;
}
ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) ". ";
  counter-increment: fourth;
}
ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) ". ";
  counter-increment: fifth;
}
ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) ". ";
  counter-increment: sixth;
}
ol > ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) "." counter(seventh) ". ";
  counter-increment: seventh;
}
ol > ol > ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) "." counter(seventh) "." counter(eigth) ". ";
  counter-increment: eigth;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) "." counter(seventh) "." counter(eigth) "." counter(ninth) ". ";
  counter-increment: ninth;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) "." counter(seventh) "." counter(eigth) "." counter(ninth) "." counter(tenth) ". ";
  counter-increment: tenth;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) "." counter(seventh) "." counter(eigth) "." counter(ninth) "." counter(tenth) "." counter(eleventh) ". ";
  counter-increment: eleventh;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) "." counter(seventh) "." counter(eigth) "." counter(ninth) "." counter(tenth) "." counter(eleventh) "." counter(twelth) ". ";
  counter-increment: twelth;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) "." counter(seventh) "." counter(eigth) "." counter(ninth) "." counter(tenth) "." counter(eleventh) "." counter(twelth) "." counter(thirteenth) ". ";
  counter-increment: thirteenth;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) "." counter(seventh) "." counter(eigth) "." counter(ninth) "." counter(tenth) "." counter(eleventh) "." counter(twelth) "." counter(thirteenth) "." counter(fourteenth) ". ";
  counter-increment: fourteenth;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) "." counter(seventh) "." counter(eigth) "." counter(ninth) "." counter(tenth) "." counter(eleventh) "." counter(twelth) "." counter(thirteenth) "." counter(fourteenth) "." counter(fifteenth) ". ";
  counter-increment: fifteenth;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > li:before {
  content: counter(first) "." counter(second) "." counter(third) "." counter(fourth) "." counter(fifth) "." counter(sixth) "." counter(seventh) "." counter(eigth) "." counter(ninth) "." counter(tenth) "." counter(eleventh) "." counter(twelth) "." counter(thirteenth) "." counter(fourteenth) "." counter(fifteenth) "." counter(sixthteenth) ". ";
  counter-increment: sixthteenth;
}
ol {
  text-indent: 0px;
}
ol > ol {
  text-indent: 10px;
}
ol > ol > ol {
  text-indent: 20px;
}
ol > ol > ol > ol {
  text-indent: 30px;
}
ol > ol > ol > ol > ol {
  text-indent: 40px;
}
ol > ol > ol > ol > ol > ol {
  text-indent: 50px;
}
ol > ol > ol > ol > ol > ol > ol {
  text-indent: 60px;
}
ol > ol > ol > ol > ol > ol > ol > ol {
  text-indent: 70px;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol {
  text-indent: 80px;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol {
  text-indent: 90px;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol {
  text-indent: 100px;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol {
  text-indent: 110px;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol {
  text-indent: 120px;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol {
  text-indent: 130px;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol {
  text-indent: 140px;
}
ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol > ol {
  text-indent: 150px;
}

</style>
</head>
<body>
<strong>Phoethe rocks</strong><br><strong>Welcome to the TGen Software Carpenty Workshop!</strong><br><br>June 19-20, 2017<br>Instructors: Nick Banovich and Emily Davenport<br>Helpers: Christophe Legendre, Elizabeth Hutchins, Eric Alsop, Ryan Richholt<br><br>Helpful links:<br>&nbsp;&nbsp;&nbsp; Workshop website: <a href="https&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;2017&#x2d;06&#x2d;19&#x2d;tgen&#x2F;">https:&#x2F;&#x2F;erdavenport.github.io&#x2F;2017-06-19-tgen&#x2F;</a><br>&nbsp;&nbsp;&nbsp;&nbsp;<br><br>Hiiiiiiii everyon<br>California<br>Phoenix<br>Where you&#x27;r<br>Tempe Subic Bay, PI<br>Phoenix<br>Georgialsfor&nbsp;<br>Washington<br>San Antonio<br>Switzerland<br>Phoenix<br>where you from?&nbsp;<br>Wisconsin<br>pain<br>Scottsdale<br>Scottsdale<br><strong>Peoria</strong><br>Phoenix<br>Long Island<br>Alabama&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; SLCma<br>Pakistan<br>ScottsdaleNJ<br>Gilbert<br>Everywhere<br>Toronto<br>Phoenix<br>Irvine, CA<br>PShoenixcottsdale<br>Utah<br>NY<br>North Carolina<br>Thunder Bay, Canada<br>Chandler<br>Ahwatukee<br>5th Floor<br>hellfire<br>Phoenix<br>Downstairs<br>az<br>what&#x27;s an imputation?<br>You can &quot;Impute&quot; what a genotype might be based on a dataset, usually if you&#x27;re missing data for a certain sample&#x2F;patient<br>what does pwd actually do?<br><br><br><br><strong># The Unix Shell (Monday morning)</strong><br><br># Go to the the unix shell lessons <a href="http&#x3a;&#x2F;&#x2F;swcarpentry&#x2e;github&#x2e;io&#x2F;shell&#x2d;novice&#x2F;">http:&#x2F;&#x2F;swcarpentry.github.io&#x2F;shell-novice&#x2F;</a> -&gt; Setup tab -&gt; download the shell-novice-data.zip file to your desktop and unzip.&nbsp;<br># You should see a folder called data-shell on your desktop. Orange sticky if you don&#x27;t<br><br># We&#x27;re going to learn about shell using this mock example of Nelle Nemo<br><br># To make the command prompt all look the same between people type the following in your terminal&#x2F;gitbash<br>PS1=&#x27;$ &#x27;<br><br><br>What&#x27;s a directory?<br>-its a folder that can contain files and other folders<br>&nbsp;<br>Can cd only be used to move into folders one step below the folder you&#x27;re already in ?<br>cd can be used to move to any directory on the system. There are two strategies you can use: Relative paths or absolute paths.<br>If we have this hypothetical file structure:<br>&nbsp;&nbsp;&nbsp; &#x2F;Users&#x2F;rrichholt&#x2F;Desktop&#x2F;code&#x2F;script.sh<br>A relative path is a path in relation to your current directory, for example, if you are already working in &#x2F;Users&#x2F;rrichholt&#x2F;Desktop&#x2F;&nbsp;<br>&quot;cd code&quot; will bring you into the code folder.&nbsp;<br>Or you can use an absolute path: cd &#x2F;Users&#x2F;rrichholt&#x2F;Desktop&#x2F;code&#x2F; will bring you to the folder from anywhere on the system.&nbsp;<br>- Do you mean change to the previous directory simply use this command &#x27;cd ..&#x27; the two periods represents previous directory<br><br>On my pc, I&#x27;m opening a notepad file - should I be in a different text editor or is this okay?<br>- notepad is great&nbsp;<br><br>How do you copy text in the terminal? I can&#x27;t use ctrl + C on sections of text.<br><br>Is pwd a general way to know where you are or which directory you are in?<br>- Yes, pwd will list where in your file system you are (which directory you are in)<br><br># Some of the commands we&#x27;ve used so far:<br><br># What is your user name on your computer?<br>whoami&nbsp;<br><br># To change directories:<br>cd<br><br># What is my current location?<br>pwd<br><br># ~ is a shortcut for your home directory. If you type &quot;cd ~&quot; it will take you back to your home dirctory<br>cd ~<br><br># Let&#x27;s go into the data-shell folder (replace username with your username, might be slightly different on a PC):<br>cd &#x2F;Users&#x2F;nbanovich&#x2F;Desktop&#x2F;data-shell<br><br># Another way to do this:<br>cd ~&#x2F;Desktop&#x2F;data-shell<br><br># How can we list out the contents of the folder we&#x27;re in?<br>ls<br><br># Let&#x27;s list out the contents of the north-pacific-gyre&#x2F;2012-07-03 folder (you should be in the data-shell folder already):<br>ls north-pacific-gyre&#x2F;2012-07-13&#x2F;<br><br># tab completion!! If you start typing the name of a file, try hitting tab. If there is only one file that matches what you&#x27;ve typed so far it will automagically complete the filename.<br># If you hit the tab once and nothing happens. Hit the tab again and it will show what files match what you&#x27;ve started to type so far.&nbsp;<br><br># Try tab completion. Start to type nor and then hit tab (and you should see north-pacific-gyre fill in).<br># Then start typing 2012- (and you should see 2012-07-03).&nbsp;<br><br># We&#x27;ve finished the first part of the shell lesson. Woot! Let&#x27;s do a few challenges.&nbsp;<br># Do the absolute vs. relative path&#x27;s challenge at the bottom of this page (<a href="http&#x3a;&#x2F;&#x2F;swcarpentry&#x2e;github&#x2e;io&#x2F;shell&#x2d;novice&#x2F;02&#x2d;filedir&#x2F;&#x29;">http:&#x2F;&#x2F;swcarpentry.github.io&#x2F;shell-novice&#x2F;02-filedir&#x2F;)</a><br><br># There&#x27;s a difference between absolute vs. relative paths. Absolute paths give the specific location of a file on that specific filesystem. Relative paths are the path to a file from the location you&#x27;re in. Absolute paths will work no matter where you are on your filesystem, but there are times when relative paths can be more useful (for example, if you work on multiple different computers)<br><br># You can list out more information about files by using the -l flag<br>ls -l<br><br># You can list out the file size as human readible (bytes) by including the -h flag<br>ls -l -h<br><br># You don&#x27;t need to have two separate flags separated by a space, you can include both after the dash<br>ls -lh<br><br># If you include -G, it will include color coding for the files<br>ls -G<br><br>BREAK! Be back at 10:30<br><br>Emilys-Pro-v2:~ erdavenport$ ls -l<br>total 16<br>drwxr-xr-x&nbsp;&nbsp; 6 erdavenport&nbsp; staff&nbsp;&nbsp; 204 May&nbsp; 5 19:28 Applications<br>drwx------+ 11 erdavenport&nbsp; staff&nbsp;&nbsp; 374 Jun 19 12:54 Desktop<br><br># Let&#x27;s start the next lesson: <a href="http&#x3a;&#x2F;&#x2F;swcarpentry&#x2e;github&#x2e;io&#x2F;shell&#x2d;novice&#x2F;03&#x2d;create&#x2F;">http:&#x2F;&#x2F;swcarpentry.github.io&#x2F;shell-novice&#x2F;03-create&#x2F;</a><br><br># Make sure you&#x27;re in the data-shell directory<br>cd ~&#x2F;Desktop&#x2F;data-shell<br><br># Make sure everything looks right<br>ls -F<br><br># Let&#x27;s make a new directory for our analysis called &quot;thesis&quot;<br>mkdir thesis<br><br># mkdir is short for &quot;make directory&quot;<br><br># List out everything again and you&#x27;ll see the new folder:<br>ls -F<br><br># Let&#x27;s list out everything in the directory thesis:<br>ls -F thesis<br><br># We don&#x27;t see any files, which is what we expect to see.&nbsp;<br><br># Let&#x27;s talk about file names<br># 1. Use names that are descriptive, but not overlylong<br># 2. AVOID SPACES. Use underscores or dashes instead. Bash interprets spaces as spaces between commands. Much easier when your file contains no spaces!<br><br># Let&#x27;s move into thesis<br>cd thesis<br><br># Let&#x27;s make a file using a basic text editor called nano (or notepad)&nbsp;<br>nano draft.txt&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # If you&#x27;re a mac<br>nano notepad&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # If you&#x27;re a PC and nano isn&#x27;t working<br><br># Whenever you write a script - use a plain text editor to write the script (versus something like Word). There are fancier ones than the ones we are using today. Nick uses Sublime and I use TextWranger -&nbsp;<br><br># In your draft document, write a message to yourself:<br>Yay I finished my thesis!!!!<br><br># Once you&#x27;ve written your text, hit control and the letter x at the same time from nano. Hit &quot;y&quot; to save.&nbsp;<br><br># Now if you type ls -F, you should see this one file in your directory<br>ls -F&nbsp;<br><br># How can we delete files? rm (stands for remove)<br>rm draft.txt<br><br><strong># CAUTION! rm does NOT put things in a trashcan. If you rm something, it&#x27;s GONE. Be careful when you rm!</strong><br><br># Let&#x27;s recreate this file<br>nano draft.txt<br><br># Let&#x27;s move one directory backwards:<br>cd ..<br><br># Try to delete the directory thesis<br>rm thesis<br><br># You should see an error. This is intentional - it saves you from deleting entire directories in one swoop.<br># If you DO want to delete a directory and everything inside it, use the -r flag (which stands for recursively)<br>rm -r thesis&#x2F;<br><br># Now if you ls -F, you won&#x27;t see the thesis directory<br><br># What if you want a failsafe?&nbsp;<br><br># Let&#x27;s make another new directory called thesis<br>mkdir thesis<br><br># cd into thesis<br>cd thesis<br><br># Once again generate this file called draft.txt<br>nano draft.txt<br><br># cd back up one directory<br>cd ..<br><br># The -i flag is called the &quot;interactive&quot; flag<br>rm -r -i thesis<br><br># Hit &quot;y&quot; when it prompts you to examine the files in your directory<br># It will ask you if you&#x27;re sure you want to remove each of those files<br># For each file then you need to say yes.<br><br># Once again make these files:<br>mkdir thesis<br>cd thesis<br>nano draft.txt<br><br># Let&#x27;s learn about moving and copying files<br><br># To move a file, type mv (stands for move). Type mv, the name of the file you want to move, and the place you want it to go (or the new name you want to give it).&nbsp;<br># mv will move a file and it is also used for renaming iles.&nbsp;<br>mv draft.txt quotes.txt<br><br># If you ls, you should only see quotes.txt<br>ls<br><br># Go ahead and move back one directory<br>cd ..<br><br># Type of the following (note that dot):<br>mv thesis&#x2F;quotes.txt .<br><br># Type ls<br>ls<br><br># You should see that quotes.txt is now in the current directory, and the thesis directory is empty<br># remember that that dot . is a shortcut for the current directory<br><br># Note that the syntax for moving files is this:<br>mv file-you-wanna-move place-or-name-you-wanna-move-it-to<br><br># What if you want to make a copy? use cp. Let&#x27;s make a copy of quotes and save that in the thesis folder with the name quotations.txt<br>cp quotes.txt thesis&#x2F;quotations.txt<br><br># List out the contents of the directory<br>ls<br><br># You should see quotes.txt in your current directory.<br># If you list out the contents of thesis, you&#x27;ll see quotations.txt<br>ls thesis<br><br># Difference between mv and cp: mv will delete the original file whereas copy will create a copy of the file<br><br># You can use ls on specific files as well as directories<br>ls -l quotes.txt<br><br># You might want to do this to see the size of the file or when it was generated<br><br># Let&#x27;s remove quotes.txt<br>rm quotes.txt<br><br># If you look inside thesis though, you&#x27;ll see quotations.txt is still there<br>ls thesis<br><br># File endings: .pdf, .txt, and .doc are all file endings that help your computer interpret what a file type is and what program to use to open it.&nbsp;<br><br># Challenge: Work through the &quot;Copy with Multiple Filenames&quot; exercise at the bottom of this page <a href="http&#x3a;&#x2F;&#x2F;swcarpentry&#x2e;github&#x2e;io&#x2F;shell&#x2d;novice&#x2F;03&#x2d;create&#x2F;">http:&#x2F;&#x2F;swcarpentry.github.io&#x2F;shell-novice&#x2F;03-create&#x2F;</a>.<br># Blue sticky when you&#x27;re done, orange if you need help!<br><br># If you copy multiple files at the same time, the final argument must be the name of a directory.&nbsp;<br><br># Make sure you&#x27;re in the data-shell folder<br><br># If you want to create an empty file, you can use the command touch<br>touch my_file.txt<br><br># Look at the size of my_file.txt<br>ls -l my_file.txt<br><br># Zero bytes in this file, because this is an empty file.&nbsp;<br><br># This might seem like a dumb command, but sometimes you need to have a file existing to write to.&nbsp;<br><br># Let&#x27;s move on to the next lesson about pipes and filters: <a href="http&#x3a;&#x2F;&#x2F;swcarpentry&#x2e;github&#x2e;io&#x2F;shell&#x2d;novice&#x2F;04&#x2d;pipefilter&#x2F;">http:&#x2F;&#x2F;swcarpentry.github.io&#x2F;shell-novice&#x2F;04-pipefilter&#x2F;</a><br><br># Make sure you&#x27;re in data-shell<br>pwd<br><br># Let&#x27;s list out the contents of the directory molecules<br>ls molecules<br><br># You can see in this folder there are a whole bunch of files with the extension .pdb (protein databank)<br><br># Move into the molecules folder<br>cd molecules<br><br># Let&#x27;s learn about the very handy wildcard character *<br>wc *.pdb<br><br># wc stands for wc. It counts the characters, words, and lines in any file<br># The wildcard character will match anything. So *.pdb will match anything in the current folder that ends with *.pdb<br><br># Output anything that starts with the letter p<br>ls p*<br><br># List anything that has the letter e in the name<br>ls *e*<br><br># To count the number of lines in each file:<br>wc -l *.pdb<br><br># Note that when using a wildcard, tabs don&#x27;t work.&nbsp;<br><br># This just displays the information to the screen. What if we wanted to save this to a file? Use the &gt; to save the output of a command (what&#x27;s called the standard output) to a file.&nbsp;<br>wc -l *.pdb &gt; lengths.txt<br><br># Now list the directory and you&#x27;ll see lengths.txt<br>ls<br><br># The cat function will list out the contents of a file to the screen<br>cat lengths.txt<br><br># cat will list out everything in a file to the screen. If it&#x27;s a big file, it will list out everything. If you just want to look at a bit, you can use less<br>less lengths.txt<br><br># less will bring up one window&#x27;s worth of information to your screen.&nbsp;<br><br># Let&#x27;s sort the contents of the file:<br>sort -n lengths.txt<br><br># What if we don&#x27;t include the -n? This treats the contents as a character rather than a number. So, if you&#x27;re trying to sort a number use -n<br>sort lengths.txt<br><br>#&nbsp; Let&#x27;s sort on the number of lines and save that to another new file called sorted_lengths.txt<br>sort -n lengths.txt &gt; sorted_lengths.txt<br><br># Another useful command to look at the top bits of files is head. Head will show the top 6 lines. The -n 1 is telling head to only print out the first line.&nbsp;<br>head -n 1 sorted_lengths.txt<br><br># If we wanted to look at the top two lines:<br>head -n 2 sorted_lengths.txt<br><br># One thing to keep in mind is to try to not write out to the same file.&nbsp;<br># Try this:<br>sort -n lengths.txt &gt; lengths.txt<br><br># Now look at the contents<br>cat lengths.txt<br><br># The file is empty, because somewhere along the way something went wrong. Rather than do this, just create a new file with the results of the sort stored in it.&nbsp;<br><br># The real power of bash comes from using what are called pipes. Pipes will take the output of a command and pass it to the next command listed after the pipe<br><br>wc -l *.pdb | sort -n<br><br># This saves you the intermediate step of having to save the one file and then type another command to get the output<br><br># Let&#x27;s add more on, let&#x27;s only look at the top line of that output<br>wc -l *.pdb | sort -n | head -n 1<br><br># What happened here?<br># 1. You first counted the lines in all of the files that end in .pdb<br># 2. The line counts are then passed to sort, which sorts by the number of lines<br># 3. That sorted list is passed to head, which then returns only the first line of the sorted list.<br><br># You can see the power here: you can chain commands together in bash&nbsp;<br><br># Let&#x27;s get to north-pacific-gyre. We&#x27;ll use a relative path here. Go back one directory (..) and the to the 2012-07-03, whcih is in north-pacific-gyre<br>cd ..&#x2F;north-pacific-gyre&#x2F;2012-07-03&#x2F;<br><br># Type pwd to make sure you&#x27;re in the right place<br>pwd<br><br># Let&#x27;s count lines of all the text files:<br>wc -l *.txt<br><br># You can see there are line counts for each of the 20 or so files in this folder<br><br># Let&#x27;s look at the number of lines in the 5 smallest files<br>wc -l *.txt | sort -n | head -n 5<br><br># All of the files should have 300 lines, but this one doesn&#x27;t have 300 lines. This is a nice trick to QC your files to double check that they&#x27;re all the same size that you expect to see.&nbsp;<br><br># You probably don&#x27;t want to include this file that only has 240 lines.&nbsp;<br><br># Let&#x27;s look at the bottom few lines rather than the top to make sure there aren&#x27;t files with more lines than we expect.&nbsp;<br>wc -l *.txt | sort -n | tail -n 5<br><br># They&#x27;re all 300 lines, so that&#x27;s good.<br># However, you&#x27;ll notice that one of the file names has a Z in the name whereas the rest of the files have A or B in the filename<br><br># Let&#x27;s list out only the files with a Z in them (must be uppercase):<br>ls *Z*.txt<br><br># See there are two samples that list the Z. She forgot to include the depth. So, there are three files that Nelle probably should exclude.&nbsp;<br><br># Let&#x27;s do a challenge. Near the bottom of this page <a href="http&#x3a;&#x2F;&#x2F;swcarpentry&#x2e;github&#x2e;io&#x2F;shell&#x2d;novice&#x2F;04&#x2d;pipefilter&#x2F;">http:&#x2F;&#x2F;swcarpentry.github.io&#x2F;shell-novice&#x2F;04-pipefilter&#x2F;</a>, do the &quot;What Does &gt;&gt; Mean?&quot; challenge.<br><br><strong># Loops</strong><br><br># Let&#x27;s move on to the next lesson about loops: <a href="http&#x3a;&#x2F;&#x2F;swcarpentry&#x2e;github&#x2e;io&#x2F;shell&#x2d;novice&#x2F;0e5&#x2d;loop&#x2F;">http:&#x2F;&#x2F;swcarpentry.github.io&#x2F;shell-novice&#x2F;0e5-loop&#x2F;</a><br><br># Loops allow us to really automate things.&nbsp;<br><br># Do a pwd to see where you are. Then navigate to the creatures directory in data-shell<br>cd Desktop&#x2F;data-shell&#x2F;creatures&#x2F;<br><br># list out contents in the creatures directory<br>ls<br><br># See that these are .dat files.&nbsp;<br><br># Let&#x27;s make some copies of these files, because we&#x27;re going to manipulate these files and we want a copy that we know hasn&#x27;t been messed with. This won&#x27;t work though, because we can&#x27;t copy multiple files.&nbsp;<br>cp *.dat original_*.dat<br><br># You can use your up arrow to scroll through commands you&#x27;ve recently used.&nbsp;<br><br># If you hit control c, it will kill the line you&#x27;re on<br><br># Control r will do a reverse search. You can search through your recent commands for things you might have done. Start typing something that was in your command (a filename for example). Keep hitting control r to scroll through the options.<br><br># Back to copying files. Let&#x27;s learn how to use a for loop to copy the files. Let&#x27;s use it to just list the head of a file rather than copy, as we&#x27;re learning<br>for filename in basilisk.dat unicorn.dat<br><br># When you hit enter, you see an arrow rather than the normal dollar sign prompt. This is because bash is waiting for you to give it more instructions. Next, type do<br>do<br>head -n 3 $filename<br>done<br><br># You can see that for each file you listed after the for, it&#x27;s listed out to the screen the top 3 lines.&nbsp;<br><br># Type echo $filename<br><br># The $ indicates that what you&#x27;re typing is a variable. After you ran the for loop, the filename unicorn.dat was saved to the variable $filename.&nbsp;<br># You can use whatever word you want as the variablename<br><br># Let&#x27;s try another loop where it prints the filename as it goes through the loop.&nbsp;<br>for filename in basilisk.dat unicorn.dat<br>do<br>echo $filename<br>done<br><br># Protip: control a will take you to the beginning of the line and control e will take you to the end of the line<br><br># Change your for loop variable to i, but don&#x27;t change the body:<br>for i in basilisk.dat unicorn.dat<br>do<br>echo $filename<br>done<br><br># You can see unicorn.dat is printed twice, because we didn&#x27;t update the variable in the body.<br><br># You can use the wildcard character to match all of the files in the directory that end in .dat<br>for filename in *.dat<br>do&nbsp;<br>head -n 3 $filename<br>done<br><br># Let&#x27;s say we want lines 80-100 for each of the files that end in .dat:<br>for filename in *.dat<br>do<br>echo $filename&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # This line prints the name of the file to the screen.<br>head -n 100 $filename | tail -n 20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # This line first takes the top 100 lines of the file, passes that with a pipe to the tail command, which takes the last 20 lines of the top 100 lines.<br>done<br><br># Back to copying multiple files: Let&#x27;s copy each of those files, renaming them as original_$filename&nbsp;<br>For filename in *.dat<br>do<br>cp $filename original_$filename<br>done<br><br># If you ls, you&#x27;ll now see the original files and then the original_$filename versions.&nbsp;<br><br># Let&#x27;s move back to the north-pacific-gyre folder<br>cd ~&#x2F;Desktop&#x2F;data-shell&#x2F;north-pacific-gyre&#x2F;2012-07-03&#x2F;<br><br># Remember we had those two files where there was a Z in the filename rather than an A or B? We don&#x27;t want to work with those files. Let&#x27;s write a loop to only look at files where there is an A or B in the filename<br>&nbsp;&nbsp;&nbsp;&nbsp;<br># We want to match any file that has either and A or B in the filename.&nbsp;<br>for datafile in *[AB].txt<br>do<br>echo $datafile<br>done<br><br># As a sanity check, you can print out anything matching A or Z<br>ls *[AZ].txt<br><br># The brackets are saying pick one letter inside these brackets to match. Can do all single digit numbers&nbsp;<br>ls *[0-9].txt<br><br>#&quot;regex&quot; is short for regular expression.<br><br># Let&#x27;s add something else after our echo command. This is demonstrating what the for loop is going to do, once we actually add a new command.&nbsp;<br>for datafile in *[AB].txt<br>do<br>echo $datafile stats-$datafile<br>done<br><br># Let&#x27;s run the stats script called goostats on each of the files. It takes in the first file after the command ($datafile) and will save the output to the second file listed after the command (stats-$datafile)<br>for datafile in *[AB].txt<br>do<br>bash goostats $datafile stats-$datafile<br>done<br><br># You can see this is taking a minute, because it&#x27;s running some computation.&nbsp;<br><br># If you type ls, now you&#x27;ll see a whole bunch of stats-*.txt files<br><br># Let&#x27;s look at the top of one of these files and you&#x27;ll see some of the stats that were generated by the goostats program<br>head stats-NENE02043B.txt<br><br># Let&#x27;s rerun that loop but make sure it&#x27;s printing out the name of the file we&#x27;re on to the screen.&nbsp;<br>for datafile in *[AB].txt<br>do<br>echo $datafile<br>bash goostats $datafile stats-$datafile<br>done<br><br># Control c will cancel an operation in bash.&nbsp;<br><br># Let&#x27;s remove anything with the stats name<br>rm stats-*<br><br># Now if you type ls, you only have the original files<br><br># Let&#x27;s write this loop one more time<br>for datafile in *[AB].txt<br>do<br>echo $datafile&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # This will print the name of the file we&#x27;re on to the screen, so we can track where we are<br>bash goostats $datafile stats-$datafile&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # This tells the computer to run the program goostats on the file $datafile, and save the output to stats-$datafile<br>done<br><br># So, you can see that we&#x27;re iterating through each one of these files and running the program on each one. This is much faster than if we were running this one by one by typing out everything each time!<br><br># Let&#x27;s do the exercise &quot;Variables in Loops&quot; in this lesson <a href="http&#x3a;&#x2F;&#x2F;swcarpentry&#x2e;github&#x2e;io&#x2F;shell&#x2d;novice&#x2F;05&#x2d;loop&#x2F;">http:&#x2F;&#x2F;swcarpentry.github.io&#x2F;shell-novice&#x2F;05-loop&#x2F;</a><br><br># Also try the exercises &quot;Saving to a File in a Loop - Part One&quot; and do Part Two as well.&nbsp;<br><br># A note about when you write your own shell scripts. You need to tell the computer that a file is executable. Nick is going to write a short little script and make it executable.<br><br># In nano, write a hello world script<br>nano helo_world.sh<br><br># write in that script<br>echo Hello World<br><br># If you type helo_world.sh, it will not execute<br><br># Let&#x27;s rename the script<br>mv helo_world.sh hello_world.sh<br><br># We can use the function chmod to change permissions. -x will make it executable.<br>chmod +x hello_world.sh<br><br># Let&#x27;s start on the finding things lesson, but the only thing we&#x27;re going to focus on is grep <a href="http&#x3a;&#x2F;&#x2F;swcarpentry&#x2e;github&#x2e;io&#x2F;shell&#x2d;novice&#x2F;07&#x2d;find&#x2F;">http:&#x2F;&#x2F;swcarpentry.github.io&#x2F;shell-novice&#x2F;07-find&#x2F;</a><br><br># Go to the writing directory:<br>cd ..&#x2F;..&#x2F;writing&#x2F;<br><br># Let&#x27;s look in haiku.txt<br>cat haiku.txt<br><br># grep is one of the most useful commands in bash. It will pull out any lines containing the word you give it<br>grep not haiku.txt<br><br># Let&#x27;s try to grep the world The<br>grep The haiku.txt<br><br># That matches both The and the beginning of Thesis, because this matches The. If you use the -w option, it will only match words<br>grep -w The haiku.txt<br><br># There are tons of options in grep. Check them out, because they can be super useful.&nbsp;<br><br># You can use wildcards with grep:<br>grep T*&nbsp;<br><br># You can do the opposite using -v and retain any lines that DON&#x27;T have The<br>grep The -v haiku.txt<br><br># If you want to ignore case you can use -i<br><br># Can save the output of grep to a file using that redirect operator &quot;&gt;&quot;. Now you can use this file in another analysis.&nbsp;<br>grep The -w haiku.txt &gt; The_haiku.txt<br><br><br><strong># R!!</strong><br><br># Go ahead and open up RStudio.<br><br># Let&#x27;s start a new project by going to File -&gt; New Project -&gt; Empty file . Save as data-carpentry<br><br># Start a new script by going to File -&gt; New File -&gt; R Script<br><br># In your finder&#x2F;file explorer, you should see a new folder called data-carpentry saved to your Desktop.&nbsp;<br><br># Go to the workshop website for the R lessons: <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;gith1ub&#x2e;io&#x2F;R&#x2d;ecology&#x2d;lesson&#x2F;index&#x2e;html">http:&#x2F;&#x2F;erdavenport.gith1ub.io&#x2F;R-ecology-lesson&#x2F;index.html</a><br><br># All the way to the right, go to the code-handout section at the top bar. and save that file.&nbsp;<br><br># Move that file to the new directory we created (data-carpentry) and rename the file data-carpentry-script.R<br><br># In RStudio, to the bottom right, make sure you&#x27;re clicked on the tab that says &quot;File&quot;. Click on the button to create a new folder, and create a new folder called data<br><br># When thinking about best practices for organizing your files, useful reference:<br><a href="http&#x3a;&#x2F;&#x2F;journals&#x2e;plos&#x2e;org&#x2F;plosbiology&#x2F;article&#x3f;id&#x3d;10&#x2e;1371&#x2F;journal&#x2e;pbio&#x2e;1001745">http:&#x2F;&#x2F;journals.plos.org&#x2F;plosbiology&#x2F;article?id=10.1371&#x2F;journal.pbio.1001745</a><br><br># How to work with RStudio:<br># The top left pane is the script we&#x27;re writing. If you hit enter, the command is not run.&nbsp;<br><br># Let&#x27;s send lines from our script to the comand line:<br>1 + 1<br><br># You can either click control return (mac or PC) or you can click on run at the top of that pane.&nbsp;<br><br># You can code directly into the R console, but keep in mind that you can&#x27;t save those commands for later. If you write commands in the script, you can save that file.&nbsp;<br><br># You can do basic math in R<br>3 + 7<br><br># Division<br>12 &#x2F; 2<br><br># Multiplication<br>5 * 5<br><br># Often we want to save values for use later. We can assign variables in R to store any info we want. Let&#x27;s store 55 in the variable weight_kg<br>weight_kg &lt;- 55<br><br># If we look in the upper right pane, you can see that there&#x27;s a variable listed.<br><br># If you type weight_kg, you can print the stored value to the screen. Note that tab completion works in RStudio!<br><br># If you want to assign to a variable and also print to the screen, you can put the assignment in ()<br>(weight_kg &lt;- 55)<br><br># Good practice is to use comments to document your code. Anything after the pound sign on a line will not be evaluated by R<br>weight_kg &lt;- 55&nbsp;&nbsp;&nbsp;&nbsp; # Assign value to weight_kg<br><br># Use comments! Future you will be happy you did because you WON&#x27;T remember why you were doing something.&nbsp;<br><br># You can now use variables to do math, because R will use whatever value has been stored in the variable to do math<br>weight_lb &lt;- weight_kg * 2.2<br><br># What happens now when we change the value in weight_kg?<br>weight_kg &lt;- 100<br><br># Let&#x27;s do another conversion and you can see that the value of weight_kg has been updated<br>weight_kg * 2.2<br><br># Math is great, but R also comes with a bunch of built in functions for doing more complex things.&nbsp;<br>square_root &lt;- sqrt(10)<br><br># Let&#x27;s try running that function on a variable that we&#x27;ve set:<br>sqrt(weight_kg)<br><br># Some functions have multiple arguments. Let&#x27;s look at the function to round numbers:<br>round(3.14159)<br><br># Let&#x27;s look at the arguments for that function though:<br>args(round)<br><br># Let&#x27;s set the number of digits to be more<br>round(3.14159, digits = 1)<br><br># If you input your argument values in the same order as shown in args(), you don&#x27;t need to write the anme of the argument:<br>round(3.14159, 1)&nbsp;&nbsp;&nbsp;&nbsp; # good!<br>round(1, 3.14159)&nbsp;&nbsp;&nbsp;&nbsp; # bad :(<br><br># If you write the argument names, you can put them in whatever order you want.&nbsp;<br>round(digis = 1, x = 3.14159)&nbsp;&nbsp;&nbsp; # Also good<br><br># Almost always, the data you input is what is listed first.&nbsp;<br><br># You can store more than one value in a variable. Let&#x27;s make what&#x27;s called a vector:<br>weight_g &lt;- c(50, 60, 65, 82)<br>weight_g<br><br># You can also include character vectors in addition to numbers<br>animals &lt;- c(&quot;mouse&quot;, &quot;rat&quot;, &quot;dog&quot;)<br>animals<br><br># What happens if you don&#x27;t put your word in quotes? R thinks it&#x27;s a variable and will throw an error if that variable isn&#x27;t already assigned.<br>animals &lt;- c(&quot;mouse&quot;, rat)<br><br># Vectors are the work horses of R. Everything in R is vectorized. We&#x27;ll see that as we learn more.&nbsp;<br><br># What are some things we might want to know about a vector?&nbsp;<br># What&#x27;s the length of my vector (how many values am I storing)?<br>length(weight_g)<br>length(animals)<br><br># Vectors in R get assigned classes (numeric, character, logical, etc). These encode information differently depending on whether something is a number or letters<br>class(weight_g)<br>class(animals)<br><br># We can use the function str() to look at the structure of a vector. Shows the class of the vector, how many variables, and the first few values.<br>str(weight_g)<br><br># chr stands for character in R, not chromosome :)<br><br>weight_g &lt;- c(weight_g, 90)<br>weight_g<br>weight_g &lt;- c(30, weight_g)<br><br># Challenge time! Do the challenge listed under &quot;Vectors and data types&quot;.&nbsp;<br><br># Subsetting vectors:<br><br># Sometimes we don&#x27;t want all of the values stored in a vector. We can pick certain ones.<br>animals &lt;- c(&quot;mouse&quot;, &quot;rat&quot;, &quot;dog&quot;, &quot;cat&quot;)<br><br># What if we only want the second value stored in our vector? Use brackets<br>animals[2]<br><br># For those of you that code in other languages, R starts numbers with 1 rather than 0!<br><br># What if we want the second and third element?<br>animals[c(2,3)]<br><br># What about the second through fourth element?<br>animals[2:4]<br><br># Let&#x27;s create a new vector made from our old vector:<br>more_animals &lt;- animals[c(1,2,3,2,1,4)]<br>more_animals<br><br># Let&#x27;s create a new vector with a bunch of numbers:<br>weight_g &lt;- c(21, 34, 54, 55, 75)<br><br># We can subset out using logicals (TRUE T and FALSE F)<br>weight_g[c(TRUE, T, F, T, F)]<br><br># Why is that helpful? Can use it to filter. For example, let&#x27;s only return the values where the weight is greater than 50<br>weight_g &gt; 50<br><br># This returns FALSE FALSE TRUE TRUE TRUE&nbsp;<br><br># Let&#x27;s combine that with subsetting to return only the values &gt; 50:<br>big_weight &lt;- weight_g[weight_g &gt; 50]<br>big_weight<br><br># What about if we want to find something equal to something else? Use two = for comparison (==).&nbsp;<br>weight_g[weight_g == 5]<br><br># You can use more than one condition when doing this type of indexing. What if we want any values where the weight is less than 30 OR less than 50? We can use the pipe |, which in R means OR<br>weight_g[weight_g &lt; 30 | weight_g &gt; 50]<br><br># You can also use and statements with &amp; (&lt;= is less than or equal to)<br>weight_g[weight_g &lt;= 30 &amp; weight_g == 21]<br><br>animals[animals == &quot;cat&quot;]<br><br># What if you have a bunch of values that you wanted to see if something was in that list of variables? Can use %in%<br>animals %in% c(&quot;rat&quot;, &quot;cat&quot;, &quot;dog&quot;, &quot;duck&quot;, &quot;goat&quot;)<br><br># Can use this to subset the original vector of animals<br>animals[animals %in% c(&quot;rat&quot;, &quot;cat&quot;, &quot;dog&quot;, &quot;duck&quot;, &quot;goat&quot;)]<br><br># Often you&#x27;ll have missing data. In R, missing data is encoded by NA<br>heights &lt;- c(2,4,4,NA,6)<br><br># Let&#x27;s try running a function like mean or max on heights<br>mean(heights)<br>max(heights)<br><br># That doesn&#x27;t work. We need to specify in an argument that we should remove NAs<br>mean(heights, na.rm = T)<br>max(heights, na.rm = T)<br><br># The exclamation point in R means is not. This returns the vector without anywhere where there was an NA<br>heights[!is.na(heights)]<br><br># Another way to remove NAs is to use complete.cases()<br>hegiths[complete.cases(heights)]<br><br># Challenge! At the very bottom of this page <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;R&#x2d;ecology&#x2d;lesson&#x2F;01&#x2d;intro&#x2d;to&#x2d;r&#x2e;html&#x23;challenge8">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;R-ecology-lesson&#x2F;01-intro-to-r.html#challenge8</a><br><br>There was a question over the break of websites you can use if you want to learn more. There are a bunch of websites where you can learn interactively. There will be material you can read and then there will be a prompt on the website where actually test those skills. The one I&#x27;ve heard about the most is datacamp for learning R or Python:<br><br><a href="https&#x3a;&#x2F;&#x2F;www&#x2e;datacamp&#x2e;com">https:&#x2F;&#x2F;www.datacamp.com</a><br><br>And there are options for learning command line as well:<br><a href="https&#x3a;&#x2F;&#x2F;www&#x2e;codecademy&#x2e;com&#x2F;en&#x2F;learn&#x2F;learn&#x2d;the&#x2d;command&#x2d;line">https:&#x2F;&#x2F;www.codecademy.com&#x2F;en&#x2F;learn&#x2F;learn-the-command-line</a><br><a href="http&#x3a;&#x2F;&#x2F;www&#x2e;learnshell&#x2e;org&#x2F;">http:&#x2F;&#x2F;www.learnshell.org&#x2F;</a><br><br># Let&#x27;s start working with data<br>download.file(&quot;<a href="https&#x3a;&#x2F;&#x2F;ndownloader&#x2e;figshare&#x2e;com&#x2F;files&#x2F;2292169">https:&#x2F;&#x2F;ndownloader.figshare.com&#x2F;files&#x2F;2292169</a>&quot;, &quot;data&#x2F;portal_data_joined.csv&quot;)<br><br># In the bottom right pane where it says files, you should now see this portal_data_joined.csv file in your file system.&nbsp;<br><br># We&#x27;ve downloaded the data, but now let&#x27;s read it into R and save it into a variable named surveys<br>surveys &lt;- read.csv(&quot;data&#x2F;portal_data_joined.csv&quot;)<br><br># We can look at the top few lines of this data.frame using head (just like in bash!)<br>head(surveys)<br><br># A data.frame is a bunch of vectors all smooshed together into columns. We can use str() to get the info for each column<br>str(surveys)<br><br># When you read in data, R looks at your data and guesses what kind of data it is.&nbsp;<br><br># We can look to see how many rows and columns our data.frame is. Row is always listed first, and columns are listed second<br>dim(surveys)<br><br># If you just want the number of rows, you can use the function nrow() (ncol() for number of columns)<br>nrow(surveys)<br>ncol(surveys)<br><br># Can look at the first few lines and can specify the number of rows to look at (let&#x27;s look at 5)<br>head(surveys, 5)<br><br># Can look at the last few lines as well using tail<br>tail(surveys, 5)<br><br># You can see the column names by typing names<br>names(surveys)<br><br># Look at the row names by using rownames(surveys)<br>rownames(surveys)<br><br># We can index and subset data.frames just like we did with vectors. What if we want the whole first row?<br>surveys[1, ]<br><br># What if we want the whole first column?<br>surveys[,1]<br><br># What if we want the value in the first row and the first column<br>surveys[1,1]<br><br># What if we want the first three rows, and the third through 5th columns?<br>surveys[1:3, 3:5]<br><br># Can use indexing to exclude rows as well. Let&#x27;s print out the first row and everything but the first column<br>surveys[1, -1]<br><br># What about the first row with everything but the second column<br>surveys[1, -2]<br><br># Note, don&#x27;t put commas in numbers in R!<br><br># What about if wanted to print just the first 6 lines, but using the -?<br>surveys[-c(7:34786), ]<br><br># You can also use column names to subset columns. Let&#x27;s pull out all rows, but just the column species_id<br>surveys[ , &quot;species_id&quot;]<br><br># Can also pick out columns by using the dollar sign. Let&#x27;s do the same as above, but using the dollar sign<br>surveys$species_id<br><br># Challenge! Let&#x27;s do the challenge here listed under &quot;Indexing and subsetting data frames&quot; <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;R&#x2d;ecology&#x2d;lesson&#x2F;02&#x2d;starting&#x2d;with&#x2d;data&#x2e;html&#x23;challenge7">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;R-ecology-lesson&#x2F;02-starting-with-data.html#challenge7</a><br><br>surveys_200 &lt;- surveys[200, ]<br><br>nrow(surveys)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # Get the number of rows in survey<br>surveys[nrow(surveys), ]&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # Pull just the last row of the dataset&nbsp;<br>surveys[nrow(surveys)&#x2F;2, ]&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # Pull the row in the middle&nbsp;<br>surveys[ , nrow(surveys)]&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # Will return an error because there aren&#x27;t that many columns<br><br># Factors - special data.class in R that deals with categorical variables.&nbsp;<br>str(surveys)<br><br># If you look at the sex column, you can see there are three levels for sex: male, female, and &quot;&quot;<br><br># Let&#x27;s create a new variable that contains female and male as factors<br>sex &lt;- factor(c(&quot;male&quot;, &quot;female&quot;, &quot;female&quot;, &quot;male&quot;))<br><br># levels() shows what the categories are (know as levels in R)<br>levels(sex)<br><br># How many levels are there? Can use nlevels()<br>nlevels(sex)<br><br># The default in R is that levels are stored alphabetically. It doesn&#x27;t sort levels in order that they are in the dataframe, but by the alphabet.&nbsp;<br># Sometimes you may want to switch the order of levels (for plotting, for example).&nbsp;<br># Can use the levels argument of factor to reset the order<br>sex &lt;- factor(sex, levels = c(&quot;male&quot;, &quot;female&quot;))<br><br># We can convert factors back to characters using the as.character() function<br>as.character(sex)<br><br># What happens with numbers though when we try to convert from factors to numbers?<br>f &lt;- factor(c(1990, 1983, 1998, 1990))<br><br># What if we just use as.numeric()<br>as.numeric(f)<br><br># AH, bad! This is an R quirk. R stores factors as numbers. When you try to convert, convert first to a character.&nbsp;<br>as.numeric(as.character(f))<br><br># This is really annoying of R and we all hate it! But be very careful when converting from factors.&nbsp;<br><br># Another way you can do this<br>as.numeric(levels(f))[f]<br><br># Let&#x27;s make our first plot<br>plot(surveys$sex)<br><br># You can see there are a whole bunch of values that are neither male nor female<br>sex &lt;- surveys$sex<br>levels(sex)<br><br># You can see there&#x27;s this random slot of missing values.&nbsp;<br># Let&#x27;s index on this first level and rename it to &quot;missing&quot;<br>levels(sex)[1] &lt;- &quot;missing&quot;<br><br># If you look at levels(sex) you can see &quot;&quot; is now &quot;missing&quot;<br>levels(sex)<br><br># Let&#x27;s replot and you can see these things have been relabeled<br>plot(surveys$sex)<br><br># Let&#x27;s read in surveys again and let&#x27;s use the argument stringsAsFactors = T<br>surveys &lt;- read.csv(&quot;data&#x2F;portal_data_joined.csv&quot;, stringsAsFactors = T)<br><br># This looks the same, but let&#x27;s set stringsAsFactors = F. This will force R to set the class as character rather than factors.<br>surveys &lt;- read.csv(&quot;data&#x2F;portal_data_joined.csv&quot;, stringsAsFactors = F)<br><br># Nick (and I do too) loads in data with stringsAsFactors = F by default.<br><br># Challenge! Do this challenge: <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;R&#x2d;ecology&#x2d;lesson&#x2F;02&#x2d;starting&#x2d;with&#x2d;data&#x2e;html&#x23;challenge13">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;R-ecology-lesson&#x2F;02-starting-with-data.html#challenge13</a><br><br># To download the tidyverse packages, in R<br>install.packages(&quot;tidyverse&quot;)<br><br># Here is the history of my commands:<br><a href="https&#x3a;&#x2F;&#x2F;www&#x2e;dropbox&#x2e;com&#x2F;s&#x2F;jo4afzs7ig4tsbb&#x2F;gHEADit&#x5f;commands&#x2e;txt&#x3f;dl&#x3d;0">https:&#x2F;&#x2F;www.dropbox.com&#x2F;s&#x2F;jo4afzs7ig4tsbb&#x2F;gHEADit_commands.txt?dl=0</a><br><br><br><br>Hi guys Nick here taking notes!<br><br>Lets get started with git!<br><br>All the problems you have with writing papers - eg writing re-writing changes - also happen when writing code. We should have a better way to track these changes.<br><br># Sets username to your name<br>git config --global user.name &quot;Your Name Here&quot;<br><br># Sets email to your email<br>git config --global user.email &quot;imsocoll@coomail.com&quot;<br><br># Makes colors consistent<br>git config --global color.ui &quot;auto&quot;<br><br># Sets the text editor you want to use<br>git config --global core.editor &quot;nano&#x2F;notepad&quot;<br><br># Look at all the settings&nbsp;<br>git config --list<br><br># Okay git will only track changes in a directory we specific so lets create a directory - we obviously don&#x27;t want to track changes of every file and folder on our computer<br>mkdir planets<br><br># I&#x27;m horrible at spelling so feel free to edit any mistakes!!!!!! Remeber levles ;)<br><br># lets move into the directory planets and check we are there<br>cd planets<br>pwd<br><br># Now we need to tell git that we want to track all of the changes within this directory. Git will tack things on a directory by directory level. Lets tell git we want to initialize a repository in the directory planets<br>git init<br><br># If we type ls we don&#x27;t see anything different<br>ls<br><br># But if we type ls -a we can see there is a hidden folder .git this folder is what makes git work.<br>ls -a<br><br>#Some people have a new prompt set. See this: <a href="https&#x3a;&#x2F;&#x2F;stackoverflow&#x2e;com&#x2F;questions&#x2F;32356595&#x2F;why&#x2d;is&#x2d;mingw64&#x2d;appearing&#x2d;on&#x2d;my&#x2d;git&#x2d;bash">https:&#x2F;&#x2F;stackoverflow.com&#x2F;questions&#x2F;32356595&#x2F;why-is-mingw64-appearing-on-my-git-bash</a><br><br># You can always change your prompt back to a simple $ using<br>PS1=&#x27;$ &#x27;<br><br># We can look at the status of our git repo using git status we will use this over and over and over and over ... you&#x27;ll see how it&#x27;s useful later today<br><br>git status<br><br># Lets do a challenge <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;git&#x2d;lessons&#x2F;03&#x2d;create&#x2e;html">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;git-lessons&#x2F;03-create.html</a><br><br># Okay lets make a new file&nbsp;<br>nano mars.txt<br><br># Lets write &quot;Mars is cold and dry, but everything is my fav color&quot; then save<br><br># Lets see whats going on with git<br><br>git status # we see that there is an untracked files &quot;mars.txt&quot;<br><br># We need to tell git to track this file lets do that now<br>git add mars.txt<br><br># Now we need to actually commit these changes. Emily gave the example of git add is telling people to line up for a picture and git commit is actually taking the picture. We always have to add a message when we commit. These should be short and descriptive.<br>git commit -m &quot;Started notes on Mars as a base&quot; # The -m tells what your message is going to be<br><br># Now lets do git status and see how it looks<br>git status #we see there is nothing to commit as the most recent version of all files have been commited<br><br># Okay lets see how we examine the commits and messages we have generated thus far<br>git log #you should see the author, data, time, and notes about the commit&nbsp;<br><br># Within that hidden .git directory is all of your tracked changes IF YOU DELETE THESE WILL BE GONE<br><br># Okay lets make some more edits and show how this works<br><br>nano mars.txt<br><br># Write something like &quot;The two moons may be a problem for wolfman&quot; # save and exit<br><br># Okay lets do another git status<br>git status # we should see that mars is now not being tacked&nbsp;<br><br># We can use another command to see the differences between the last commit and our new file&nbsp;<br>git diff<br><br># Lets try and commit this file<br>git commit -m &quot;Add concerns about effects of Mars&#x27; moons on Wolfman&quot;<br><br># Let&#x27;s look at what happend<br>git status&nbsp;<br><br># This didn&#x27;t work because we always have to do the three step process. So let&#x27;s try again with an add command first<br>git add mars.txt<br>git commit -m &quot;Add concerns about effects of Mars&#x27; moons on Wolfman&quot;<br><br>#Okay, let&#x27;s try and figure out more about staging. Start by making a new change<br>nano mars.txt<br># Write &quot;But the Mummy will appreciate the lack of humidity&quot; and save<br><br># Now let&#x27;s try git diff<br>git diff<br><br># Now lets see what happens after we add mars.txt<br>git add mars.txt<br>git diff # we see there are no changes to show! Once a file is staged it is being tracked by git so git diff doesn&#x27;t work.<br><br># We can look at changes in the staged files with another argument<br>git diff --staged<br><br># Now let&#x27;s commit this final change<br>git commit -m &quot;Discuss concerns about Mars&#x27; climate for Mummy&quot;<br><br># Okay let&#x27;s look at the git status to make sure everything is working and the git log which should show three commits and three messages<br>git status<br>git log<br><br># Lets do a challange while I catch up on notes! Very bottom of this page: <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;git&#x2d;lessons&#x2F;04&#x2d;changes&#x2e;html">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;git-lessons&#x2F;04-changes.html</a><br><br># Okay let&#x27;s move back to planets and get ready to explore our history! <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;git&#x2d;lessons&#x2F;05&#x2d;history&#x2e;html">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;git-lessons&#x2F;05-history.html</a><br>cd ~&#x2F;planets<br><br># Okay lets look at the current version of mars.txt with the previous version. We&#x27;ll do this using a shortcut<br><br>git diff HEAD~1<br><br># We can also do this by looking at the commit identifiers. Remeber if we type git log we can see the identifiers<br>git log # Then copy and past the identifier you want to look at - you can sometimes do the first 10-15 identifiers<br><br>git diff IDENTIFIER_YOU_COPIED mars.txt<br><br># Okay but what if we want to go back to an older version? Let start by messing up mars.txt<br>nano mars.txt # delete everything and write a bunch of junk<br><br># Oh no we just deleted all of our hard work. We can use checkout to get the last version<br>git checkout HEAD mars.txt #This checks out the last committed version of mars.txt<br># You can also use the unique commit identifier in place of head<br><br># Okay we usually don&#x27;t want to track changes to data files - espcially things like raw sequecing data. Let&#x27;s see how we&#x27;d do this. Let&#x27;s make a directory and a few fake files<br>mkdir results<br>touch a.dat b.dat c.dat results&#x2F;a.out results&#x2F;b.out<br><br># Okay if we do a git status we&#x27;ll see these files all show up as untracked<br>git status<br><br># Now let&#x27;s create a special file recognized by git called .gitignore<br>nano .gitignore<br># In here let&#x27;s write &quot;*.dat and on a new line results&#x2F;&quot; This is telling git to ignore everything that ends with .dat - remember our wildcard and everything in our results folder.<br><br># So now let&#x27;s do git status<br>git status # We can see the gitignore file shows up but none of the other data or results files<br><br># Let&#x27;s add the gitignore file for good practice<br>git add .gitignore<br>git commit -m &quot;Add the ignore file&quot;<br>git status<br><br># We can see we&#x27;re all caught up even though we aren&#x27;t tacking our data files. Let&#x27;s see what happens if we try to add one<br>git add a.dat # You should get a message saying this doesn&#x27;t work. This is good because we don&#x27;t want to do this on accent - eg acciently start tracking a 100G bam file<br><br># If you want to see what files you&#x27;re ignoring&nbsp;<br>git status --ignored<br><br># Everyone needs to go and create a new account on github.com<br><br># Okay guys time to get in deep. On the internet go to github.com and create a new repository let&#x27;s use the same name as we did on our computer - planets<br><br># Now we need to connect the repository on the internet and the command line using this code in the command line<br>git remote add orgin <a href="https&#x3a;&#x2F;&#x2F;github&#x2e;com&#x2F;USERNAME&#x2F;planets&#x2e;git">https:&#x2F;&#x2F;github.com&#x2F;USERNAME&#x2F;planets.git</a> # The https link comes from the github.com repository Use your username not mine<br><br># We can use git remote -v to look at where this repository is connected to the web<br>git remote -v<br><br># Okay so we have connected these repositotries but we haven&#x27;t synced the files on our computer to the website. Let&#x27;s do that now using git push<br>git push origin master<br><br># We can also pull information from the internet using pull<br>git pull origin master<br><br># Okay let&#x27;s simulate working on two computers. Lets first move to the tmp folder on you computer from the terminalls<br>cd &#x2F;tmp<br><br># Now we are going to clone the directory from the internet. Go to the github website and navigate to the repository. You should see a gree button that says &quot;Clone or Download&quot;. Click that and copy the link. Go back to the terminal and do the following<br>git clone LINK_FROM_GITHUB<br><br># If you do an ls you should see the planets directory<br>ls<br><br># Okay so within your &#x2F;tmp&#x2F;planets directory let&#x27;s make a new file<br>nano pluto.txt # Write something awesome<br><br># Let&#x27;s add and commit<br>git add pluto.txt<br>git commit -m &quot;Some notes about Pluto&quot;<br>cl# Okay if we do git status we see that everything is up to date, but we get a message that we are ahead of the origin&#x2F;master<br>git status<br><br># Let&#x27;s push the changes to the world wide web<br>git push origin master<br><br># You should now see that this file is on your repository online. This is still not synced on your other &quot;computer&quot; - i.e. ~&#x2F;planets&#x2F; . Lets navigate there.<br><br>cd ~&#x2F;planets&#x2F;<br><br># Let&#x27;s pull down the most recent commits from the webs<br><br>git pull origin master<br><br># Okay let&#x27;s make a change to mars.txt in ~&#x2F;planets<br>nano mars.txt<br><br># Let&#x27;s add and commit<br>git add mars.txt<br>git commit -m &quot;added a line to our home copy&quot;<br><br># Let&#x27;s push this to the internet<br>git push origin master<br><br># Okay let&#x27;s navigate to the &#x2F;tmp&#x2F;planets and make some differnt chagnes to mars.txt in this direcotry<br>nano mars.txt<br><br># Let&#x27;s add, commit, and push<br>git add mars.txt<br>git commit -m &quot;added diff line to tmp copy&quot;<br>git push origin master<br><br># This should have given you an error when you tried to push<br># Let&#x27;s go ahead and git pull<br>git pull origin master<br><br># Now if you cat mars.txt<br>cat mars.txt<br><br># You&#x27;ll see both changes are displayed here. We have to clean these up by hand. Let&#x27;s do nano and delete all of the junk and add a new fourth line saying we fixed merge.<br>nano mars.txt<br><br># Let&#x27;s git add and commit<br>git add mars.txt<br>git commit -m &quot;fixed merge issues&quot;<br><br># Let&#x27;s do git status to see what&#x27;s happening here<br>git status # everything looks good<br><br># Let&#x27;s do git push<br>git push origin master # now this works because we have fixed the errors<br><br># Okay let&#x27;s navigate back to the ~&#x2F;planets directory and do a git pull<br>cd ~&#x2F;planets<br>git pull origin master<br><br># Let&#x27;s look at mars.txt now<br>cat mars.txt<br><br># Now we can see that mars.txt is up to date everywhere. We didn&#x27;t get another merge error since we told git exactly what we wanted when we did the manual edits. git understands we did this and so no merge error1<br><br># Get geared up for a challenge!!!!<br><br># We are going to do the challange at the bottom of this page: <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;git&#x2d;lessons&#x2F;09&#x2d;conflict&#x2e;html">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;git-lessons&#x2F;09-conflict.html</a><br><br># We are going to clone a repository from emily <a href="https&#x3a;&#x2F;&#x2F;github&#x2e;com&#x2F;erdavenport&#x2F;bio&#x2e;git">https:&#x2F;&#x2F;github.com&#x2F;erdavenport&#x2F;bio.git</a>. Clone this anywhere you want as long as it&#x27;s not inside planets. &#x2F;tmp is a good option<br><br><br># Okay guys hope you had a good lunch. Get ready for R.<br><br># Alright let&#x27;s re-read in our data from yesterday<br>surveys &lt;- read.csv(&quot;data&#x2F;portal_data_joined.csv&quot;)<br><br># Let&#x27;s look at the first few lines of the data frame to see how everythging looks<br>head(surveys)<br><br># Let&#x27;s do a quick plot of our data<br><br>hist(surveys$weights) # hist is a function in R which allows you to make histograms<br><br># Remember to code in the source (top left) and send to console (bottom left) using command or control + enter<br><br># We can change the color of the plot using the col argument within hist and the title using main<br>hist(sruveys$weight, col = &quot;red, main = &quot;Distribution of weights&quot;) # the color and title need to be within double quotes<br><br># Alright get ready for some loops!<br><br>#First we are going to do use the command if<br>if(surveys$year[1] == 1984){ # This is saying is the first element of the survey$year vector is euqal to 1984 then advance<br><ul class="indent"><li>print(&quot;Great Scott, it&#x27;s 1984&quot;) # You&#x27;ll notice this is positioned to the right by one tab - this is common synatx for loops. Anything after the { is usually indented to say this is within the loop</li></ul>}<br><br># This didn&#x27;t print anything since the first element is 1977. Let&#x27;s update the code with a new statement called else&nbsp;<br><br>if(surveys$year[1] == 1984){<br><ul class="indent"><li>print(&quot;Great Scott, it&#x27;s 1984&quot;)&nbsp;</li><li>} else { # This else statement tells R to do something in the case where the if statment isn&#x27;t true</li><li>print (&quot;It&#x27;s not 1984)</li><li>}</li><li><br></li></ul># Challenge is here about halfway down: <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;R&#x2d;ecology&#x2d;lesson&#x2F;03&#x2d;loops&#x2d;and&#x2d;functions&#x2e;html">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;R-ecology-lesson&#x2F;03-loops-and-functions.html</a><br><br># Let&#x27;s solve the challange together!<br>if(surveys$weight[40] &gt;= 28.3){<br>&nbsp;&nbsp;&nbsp; print(&quot;It&#x27;s a biggun&quot;)<br>&nbsp;&nbsp;&nbsp; } else {&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; print (&quot;It&#x27;s not very big&quot;)<br><ul class="indent"><li>}</li><li><br></li></ul># Okay now let&#x27;s do some for loops. These are similar in principle to the for loops we learned about in bash yesterday<br>for (i in 1:10){<br>&nbsp;&nbsp;&nbsp; print(i)<br>&nbsp;&nbsp;&nbsp; }<br>#This should print 1-10. What happens if we print out a string?<br><br>for (i in 1:10{<br>&nbsp;&nbsp;&nbsp; print(&quot;cookie&quot;)<br>&nbsp;&nbsp;&nbsp; }<br>#This prints cookie 10 times because nothing is adaptive.<br><br># Let&#x27;s loop over something non-numeric<br>for(i in c(&quot;cat&quot;, &quot;dog&quot;, &quot;mouse&quot;)){<br>&nbsp;&nbsp;&nbsp; print(i)<br>&nbsp;&nbsp;&nbsp; }<br># You can see this prints cat, dog, and mouse<br><br># We can use this function to loop over the rows in surveys. Let&#x27;s try that now.<br>for( i in 1:dim(surveys)[1]){ #This indexing is saying to loop from 1 to the number of rows in surveys - ie ~37K<br>&nbsp; if (surveys$year[i] == 1984){ # This is called a nested loop as it is within the other for loop<br>&nbsp;&nbsp;&nbsp; print(&quot;Great Scott, it&#x27;s 1984&quot;)<br>&nbsp; } else {<br>&nbsp;&nbsp;&nbsp; print(&quot;It&#x27;s not 1984&quot;)<br>&nbsp; }<br>}<br><ul class="indent"><li><br></li></ul># Okay let&#x27;s do something more useful. First let&#x27;s create a new data frame<br>surveys_adjusted &lt;- surveys # This just made of copy of surveys called surveys_adjusted<br><br># Now let&#x27;s make a new loop where we cahgne the weight variable by 10% if the year is 1984<br>for( i in 1:dim(surveys_adjusted)[1]){&nbsp;<br>&nbsp; if (surveys_adjusted$year[i] == 1984){&nbsp;<br>&nbsp;&nbsp;&nbsp; surveys_adjusted$weight[i] &lt;- surveys_adjusted$weight[i] * 1.1<br>&nbsp; }&nbsp;<br><br><br>#Okay now that we succesfully built this awesome for loop let&#x27;s see if it worked!<br><br>#Let&#x27;s look at the mean weight of samples from 1984 in the adjusted and non-adjusted data frames<br>o_weights &lt;- mean(surveys$weight[surveys$year == 1984], na.rm = T)<br><br>a_weights &lt;- mean(surveys_adjusted$weight[surveys_adjusted$year == 1984], na.rm = T)<br><br>#Challange acepted: <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;R&#x2d;ecology&#x2d;lesson&#x2F;03&#x2d;loops&#x2d;and&#x2d;functions&#x2e;html&#x23;challenge5">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;R-ecology-lesson&#x2F;03-loops-and-functions.html#challenge5</a><br><br>surveys_adj_no_na &lt;- surveys_adjusted[!<strong>is.na</strong>(surveys_adjusted$weight), ]<br>big_animals &lt;- 0<br>for( i in 1:nrow(surveys_adj_no_na)){<br>&nbsp;&nbsp;&nbsp; if (surveys_adj_no_na$weight[i] &gt; 28.3 {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; big_animals &lt;- big_animals + 1<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; }<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; }<br>big animals<br><br># Time for dplyr and ggplot2: <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;R&#x2d;ecology&#x2d;lesson&#x2F;04&#x2d;dplyr&#x2e;html">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;R-ecology-lesson&#x2F;04-dplyr.html</a><br><br># Let&#x27;s load up the tidyverse (code name for dplyr plust some other fun packages to make the data tidy)<br>library(tidyverse)<br><br># Let&#x27;s subset on some columns<br>select(surveys, plot_id, species_id, weight) # similar to some of our other indexing but it doesn&#x27;t need quote<br><br># Let&#x27;s subset using a slice<br>select(surveys, month:year)<br><br># Okay now let&#x27;s try to filter by a row<br>filter(surveys, year == 1995) # This selects all rows where surveys$year is equal to 1995<br><br># Let&#x27;s explore the utility of pipes. Let&#x27;s first do an operation with pipes. First lets filter on surveys with wieght less than 5<br>weight5 &lt;- filter(surveys, weight &lt; 5)<br><br># Then select three columns<br>select(weight5, species_id, sex, weight)&nbsp;<br><br># Now let&#x27;s try this with a pipe. This remove the intermediate variable and makes things go a bit faster<br>surveys_sml &lt;- surveys %&gt;%&nbsp;<br>&nbsp; filter(weight &lt; 5) %&gt;%<br>&nbsp; select(species_id, sex, weight)<br>&nbsp;&nbsp;<br>&nbsp;&nbsp;<br>#Challenge - Using pipes, subset the survey data to include individuals collected before 1995 and retain only the columns year, sex, and weight.<br><br>surveys %&gt;%&nbsp;<br>&nbsp; filter(year &lt; 1995) %&gt;%<br>&nbsp; select(year, sex, weight)<br>&nbsp;&nbsp;<br># MUTATE - not in the cancer&#x2F;X-men sense of the word<br>surveys %&gt;%<br>&nbsp; mutate(weight_kg = weight&#x2F;1000) # This returns a new data frame with an additional column called wieght_kg. This data frame could be saved<br><br># Mutate looking at only the first few lines<br>surveys %&gt;%<br>&nbsp; mutate(weight_kg = weight&#x2F;1000) %&gt;%<br>&nbsp; head<br><br># Now let&#x27;s filter out those pesky NAs<br>surveys %&gt;%<br>&nbsp; filter(!is.na(weight)) %&gt;%<br>&nbsp; mutate(weight_kg = weight&#x2F;1000) %&gt;%<br>&nbsp; head<br>&nbsp;&nbsp;<br>&nbsp; #Challange: <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;R&#x2d;ecology&#x2d;lesson&#x2F;04&#x2d;dplyr&#x2e;html&#x23;mutate">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;R-ecology-lesson&#x2F;04-dplyr.html#mutate</a><br>&nbsp; surveys %&gt;%<br>&nbsp; filter(!is.na(hindfoot_length)) %&gt;%<br>&nbsp; mutate(hindfoot_half = hindfoot_length&#x2F;2) %&gt;%<br>&nbsp; filter(hindfoot_half &lt; 30) %&gt;%<br>&nbsp; select(species_id, hindfoot_half)<br>&nbsp;&nbsp;<br>&nbsp; # Okay, let;s see how we can use a varable to group and summarize. Here we will group by sex and calculate the mean weight for each grouping<br>&nbsp; surveys %&gt;%&nbsp;<br>&nbsp; group_by(sex) %&gt;%<br>&nbsp; summarise(mean_weight &lt;- mean(weight, na.rm =T))<br>&nbsp;&nbsp;<br>&nbsp; # We can also group by multiple categories - here we will group out by sex and species&nbsp;<br>surveys %&gt;%&nbsp;<br>&nbsp; filter(!is.na(weight)) %&gt;%<br>&nbsp; group_by(sex, species_id) %&gt;%<br>&nbsp; summarise(mean_weight &lt;- mean(weight))<br>&nbsp;&nbsp;<br># Let&#x27;s get our data ready to work in ggplot2 by removing missing values and NAs<br>surveys_complete &lt;- surveys %&gt;%<br>&nbsp; filter(species_id != &quot;&quot;, !is.na(weight), !is.na(hindfoot_length), sex != &quot;&quot;)<br><br>#If you are following along here is the link for the next lesson. Let&#x27;s get our plot on! <a href="http&#x3a;&#x2F;&#x2F;erdavenport&#x2e;github&#x2e;io&#x2F;R&#x2d;ecology&#x2d;lesson&#x2F;05&#x2d;visualization&#x2d;ggplot2&#x2e;html">http:&#x2F;&#x2F;erdavenport.github.io&#x2F;R-ecology-lesson&#x2F;05-visualization-ggplot2.html</a><br><br>#Let&#x27;s initialize a ggplot<br>ggplot(data = surveys_complete) # This should just give a grey background<br><br>#Let&#x27;s add some axis<br>ggplot(data = surveys_complete,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = weight, y = hindfoot_length)) # This should give you the specified x and y axis<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<br># Let&#x27;s actually make a scatter plot<br>ggplot(data = surveys_complete,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = weight, y = hindfoot_length)) +&nbsp;<br>&nbsp; geom_point() # geom_point is the geom for scatterplot - a geom is a shape or graph type in ggplot<br>&nbsp;&nbsp;<br># We can save this base ggplot as an object<br>surveys_plot = ggplot(data = surveys_complete,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = weight, y = hindfoot_length))<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<br># Now to get the scatter plot we can do this<br>surveys_plot + geom_point()<br><br># Some of these points are hard to see so let&#x27;s add an argument to geom_point which will shade on density<br>ggplot(data = surveys_complete,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = weight, y = hindfoot_length)) +<br>&nbsp; geom_point(alpha = 0.1)<br><br># Now let&#x27;s add some awesome color<br>ggplot(data = surveys_complete,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = weight, y = hindfoot_length)) +<br>&nbsp; geom_point(alpha = 0.1, color = &quot;blue&quot;)<br>&nbsp;&nbsp;<br># We can also add another aesthetic (aes) to color by species<br>ggplot(data = surveys_complete,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = weight, y = hindfoot_length)) +<br>&nbsp; geom_point(alpha = 0.1, aes(color = species_id))<br>&nbsp;&nbsp;<br># Okay, let&#x27;s try a slighlty differnt graph. Let&#x27;s look at species_id and hindfoot_length in a boxplot rather than a scatter plot<br>ggplot(data = surveys_complete,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = species_id, y = hindfoot_length)) +&nbsp;<br>&nbsp; geom_boxplot()<br><br># Now lets add some ponts to the boxplot where we show points with a &quot;jitter&quot;<br>ggplot(data = surveys_complete,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = species_id, y = hindfoot_length)) +&nbsp;<br>&nbsp; geom_boxplot() + geom_jitter(alpha = 0.3, color = &quot;tomato&quot;)<br>&nbsp;&nbsp;<br># The boxplot here gets a bit lost, so let&#x27;s flip jitter and boxplot<br>ggplot(data = surveys_complete,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = species_id, y = hindfoot_length)) +&nbsp;<br>&nbsp; geom_jitter(alpha = 0.3, color = &quot;tomato&quot;) + geom_boxplot()&nbsp;<br><br># Okay let&#x27;s combine some of what we already learned and ggplot<br>yearly_counts &lt;- surveys_complete %&gt;%&nbsp;<br>&nbsp; group_by(year, species_id) %&gt;%<br>&nbsp; tally # This counts for each group<br><br># Now let&#x27;s use these data to make a new plot<br>ggplot(data = yearly_counts,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = year, y = n)) +&nbsp;<br>&nbsp; geom_line()<br>&nbsp;&nbsp;<br>&nbsp; # This does not look like what we want let&#x27;s do a better job splitting by group and coloring each line<br>&nbsp; # Now let&#x27;s use these data to make a new plot<br>ggplot(data = yearly_counts,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = year, y = n, group = species_id, color = species_id)) +&nbsp;<br>&nbsp; geom_line()<br>&nbsp;&nbsp;<br><br># Let&#x27;s talk about faceting<br>ggplot(data = yearly_counts,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = year, y = n, group = species_id, color = species_id)) +&nbsp;<br>&nbsp; geom_line() +<br>&nbsp; facet_wrap(~ species_id) # this breaks each species_id into an individual plot<br>&nbsp;&nbsp;<br><br># Let&#x27;s prepare our data so we can also group by sex<br>yearly_sex_counts &lt;- surveys_complete %&gt;%&nbsp;<br>&nbsp; group_by(year, species_id, sex) %&gt;%<br>&nbsp; tally&nbsp;<br><br># Now let&#x27;s change our grouping to be by sex instead of species_id<br>ggplot(data = yearly_sex_counts,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = year, y = n, group = sex, color = sex)) +&nbsp;<br>&nbsp; geom_line() +<br>&nbsp; facet_wrap(~ species_id)<br>&nbsp;&nbsp;<br># Now let&#x27;s tweak some more of our plot parameters. We are going to get rid of the grey backgound&nbsp;<br>ggplot(data = yearly_sex_counts,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = year, y = n, group = sex, color = sex)) +&nbsp;<br>&nbsp; geom_line() +<br>&nbsp; facet_wrap(~ species_id) +<br>&nbsp; theme_bw() # get&#x27;s rid of the grey background but keeps the grid lines<br>&nbsp;&nbsp;<br>&nbsp; # Let&#x27;s save this plot<br>&nbsp; myplot &lt;- ggplot(data = yearly_sex_counts,&nbsp;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; aes(x = year, y = n, group = sex, color = sex)) +&nbsp;<br>&nbsp; geom_line() +<br>&nbsp; facet_wrap(~ species_id) +<br>&nbsp; theme_bw()<br>&nbsp; ggsave(myplot, file = &quot;mysweetplot.pdf&quot;)<br><br>
<div style="display:none"><a href="/javascript" data-jslicense="1">JavaScript license information</a></div>
</body>
</html>
